tensorboard/webapp/analytics/index.ts

```diff
@@ -0,0 +1,64 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {Action, ActionReducer} from '@ngrx/store';
+import {coreLoaded} from '../core/actions';
+import {getActivePlugin} from '../core/store';
+
+// We fake the global 'ga' object, so the object is a noop. The
+// google.analytics typing gives the object a type of UniversalAnalytics.ga.
+// We do not track open source users.
+declare const ga: Function;
+
+class EventSender {
+  static sendPageView(pathAndPlugin: string) {
+    ga('set', 'page', pathAndPlugin);
+    ga('send', 'pageview');
+  }
+}
+
+// Send analytics when the activePlugin changes.
+export function analyticsMetaReducer(
+  reducer: ActionReducer<any>
+): ActionReducer<any> {
+  // Must wait for 'core' to load before using its 'getActivePlugin' selector.
+  let isCoreLoaded = false;
+
+  return (state, action: Action) => {
+    if (action.type === coreLoaded.type) {
+      isCoreLoaded = true;
+      return reducer(state, action);
+    }
+    if (!isCoreLoaded) return reducer(state, action);
+
+    const oldActivePlugin = util.getActivePlugin(state);
```

Comment:
How about you just take the plugin string from the action payload instead? 

For instance, I really expected to see some kind of switch statement based on the action type and don't compute nextState or make this module stateful like this.

Reply:
> How about you just take the plugin string from the action payload instead?

This makes sense for `changePlugin` actions, but not really for `pluginsListingLoaded`.  The latter's payload is just a list of plugins, and only the CoreReducer is responsible for knowing how to update the `activePlugin`.  Having the analytics code know how pluginsListingLoaded selects a default plugin doesn't make sense.

The core reason for the before/after check is that anyone dispatching `changePlugin` doesn't know whether it will actually change the active plugin, so our metaReducer must check itself.  

That said, maybe it's not so bad, having a metaReducer assume that inner reducers may affect the result of a selector in some way?  (unless this is an anti-pattern or some principle is violated)

Reply:
Hm alternatively, wdyt about this:

Anyone dispatching `changePlugin` today should really be dispatching `changePluginRequested`.  We introduce a new coreEffect that says, upon `changePluginRequested`, if the requested plugin is different from the current, dispatch a `pluginChanged` action that actually modifies the state's `activePlugin`.  Then, we're guaranteed only `pluginChanged` corresponds to actual plugin changes.

Instead of having `pluginsListingLoaded` directly change the state's `activePlugin`, there should be a coreEffect that listens for `pluginsListingLoaded` and dispatches a `changePluginRequested` to set the default plugin.

?